File: internal/notifications/extraction.go
// extraction.go ports src-tauri/src/notifications/trigger_checker/extraction.rs.
// Extracts tool results from a ParsedMessage and provides tool summary labels.
package notifications

import (
	"encoding/json"
	"fmt"

	"claude-devtools/internal/domain"
)

// ExtractedToolResult mirrors trigger_checker/extraction.rs ExtractedToolResult.
type ExtractedToolResult struct {
	ToolUseID string
	IsError   bool
	Content   string
	ToolName  *string
}

// ExtractToolResults collects all tool results from a message.
// Mirrors trigger_checker/extraction.rs::extract_tool_results.
func ExtractToolResults(msg *domain.ParsedMessage) []ExtractedToolResult {
	var results []ExtractedToolResult

	// From msg.ToolResults (pre-extracted by entry_parser).
	for i := range msg.ToolResults {
		tr := &msg.ToolResults[i]
		toolName := findToolNameByID(msg, tr.ToolUseID)
		results = append(results, ExtractedToolResult{
			ToolUseID: tr.ToolUseID,
			IsError:   tr.IsError,
			Content:   valueToContentString(tr.Content),
			ToolName:  toolName,
		})
	}

	// From msg.ToolUseResult (raw tool_use_result JSON blob).
	if msg.ToolUseResult != nil {
		var obj map[string]json.RawMessage
		if json.Unmarshal(msg.ToolUseResult, &obj) == nil {
			isError := false
			if v, ok := obj["isError"]; ok {
				_ = json.Unmarshal(v, &isError)
			}
			if !isError {
				if v, ok := obj["is_error"]; ok {
					_ = json.Unmarshal(v, &isError)
				}
			}

			var toolUseID string
			if v, ok := obj["toolUseId"]; ok {
				_ = json.Unmarshal(v, &toolUseID)
			}
			if toolUseID == "" && msg.SourceToolUseID != nil {
				toolUseID = *msg.SourceToolUseID
			}

			if toolUseID != "" {
				var toolName *string
				if v, ok := obj["toolName"]; ok {
					var s string
					if json.Unmarshal(v, &s) == nil {
						toolName = &s
					}
				}
				content := extractContentFromToolUseResult(msg.ToolUseResult)
				results = append(results, ExtractedToolResult{
					ToolUseID: toolUseID,
					IsError:   isError,
					Content:   content,
					ToolName:  toolName,
				})
			}
		}
	}

	// From content blocks (tool_result entries).
	for i := range msg.Content.Blocks {
		b := &msg.Content.Blocks[i]
		if b.Type != "tool_result" || b.ToolUseID == nil || b.Content == nil {
			continue
		}
		content := toolResultContentValue(b.Content)
		isError := b.IsError != nil && *b.IsError
		toolName := findToolNameByID(msg, *b.ToolUseID)
		results = append(results, ExtractedToolResult{
			ToolUseID: *b.ToolUseID,
			IsError:   isError,
			Content:   content,
			ToolName:  toolName,
		})
	}

	return results
}

// findToolNameByID searches tool calls for the name matching a tool_use ID.
func findToolNameByID(msg *domain.ParsedMessage, toolUseID string) *string {
	for i := range msg.ToolCalls {
		if msg.ToolCalls[i].ID == toolUseID {
			name := msg.ToolCalls[i].Name
			return &name
		}
	}
	if msg.SourceToolUseID != nil && *msg.SourceToolUseID == toolUseID {
		if msg.ToolUseResult != nil {
			var obj map[string]json.RawMessage
			if json.Unmarshal(msg.ToolUseResult, &obj) == nil {
				if v, ok := obj["toolName"]; ok {
					var s string
					if json.Unmarshal(v, &s) == nil {
						return &s
					}
				}
			}
		}
	}
	return nil
}

// GetToolSummary produces a short human-readable label for a tool_use call.
// Mirrors trigger_checker/extraction.rs::get_tool_summary.
func GetToolSummary(toolName string, input json.RawMessage) string {
	getStr := func(field string) string {
		var obj map[string]json.RawMessage
		if json.Unmarshal(input, &obj) != nil {
			return ""
		}
		v, ok := obj[field]
		if !ok {
			return ""
		}
		var s string
		if json.Unmarshal(v, &s) == nil {
			return s
		}
		return ""
	}

	switch toolName {
	case "Read", "Edit", "Write":
		if fp := getStr("file_path"); fp != "" {
			// Take the last path segment.
			last := fp
			for i := len(fp) - 1; i >= 0; i-- {
				if fp[i] == '/' || fp[i] == '\\' {
					last = fp[i+1:]
					break
				}
			}
			return last
		}
		return toolName
	case "Bash":
		if cmd := getStr("command"); cmd != "" {
			if len(cmd) > 60 {
				return fmt.Sprintf("%s...", cmd[:60])
			}
			return cmd
		}
		return "shell command"
	case "Grep", "Glob":
		if p := getStr("pattern"); p != "" {
			return p
		}
		return toolName
	}
	return toolName
}

File: src-tauri/src/notifications/extraction.rs
//! Extracts tool results from a message and produces tool summary labels.
//! Ported from `internal/notifications/extraction.go`.

use serde_json::Value;

use crate::types::jsonl::ContentBlock;
use crate::types::messages::ParsedMessage;

use super::tool_maps::{
    extract_content_from_tool_use_result, message_blocks, tool_result_content_value,
    value_to_content_string,
};

#[derive(Debug, Clone)]
pub struct ExtractedToolResult {
    pub tool_use_id: String,
    pub is_error: bool,
    pub content: String,
    pub tool_name: Option<String>,
}

/// Collects all tool results from a message.
pub fn extract_tool_results(msg: &ParsedMessage) -> Vec<ExtractedToolResult> {
    let mut results = Vec::new();

    // From pre-extracted tool_results.
    for tr in &msg.tool_results {
        let tool_name = find_tool_name_by_id(msg, &tr.tool_use_id);
        results.push(ExtractedToolResult {
            tool_use_id: tr.tool_use_id.clone(),
            is_error: tr.is_error,
            content: value_to_content_string(&tr.content),
            tool_name,
        });
    }

    // From the raw tool_use_result JSON blob.
    if let Some(tur) = &msg.tool_use_result {
        if let Some(obj) = tur.as_object() {
            let mut is_error = obj.get("isError").and_then(Value::as_bool).unwrap_or(false);
            if !is_error {
                is_error = obj
                    .get("is_error")
                    .and_then(Value::as_bool)
                    .unwrap_or(false);
            }

            let mut tool_use_id = obj
                .get("toolUseId")
                .and_then(Value::as_str)
                .unwrap_or("")
                .to_string();
            if tool_use_id.is_empty() {
                if let Some(src) = &msg.source_tool_use_id {
                    tool_use_id = src.clone();
                }
            }

            if !tool_use_id.is_empty() {
                let tool_name = obj
                    .get("toolName")
                    .and_then(Value::as_str)
                    .map(str::to_string);
                let content = extract_content_from_tool_use_result(tur);
                results.push(ExtractedToolResult {
                    tool_use_id,
                    is_error,
                    content,
                    tool_name,
                });
            }
        }
    }

    // From content blocks (tool_result entries).
    for block in message_blocks(msg) {
        if let ContentBlock::ToolResult {
            tool_use_id,
            content,
            is_error,
        } = block
        {
            let tool_name = find_tool_name_by_id(msg, tool_use_id);
            results.push(ExtractedToolResult {
                tool_use_id: tool_use_id.clone(),
                is_error: is_error.unwrap_or(false),
                content: tool_result_content_value(content),
                tool_name,
            });
        }
    }

    results
}

/// Searches tool calls (then the raw result blob) for a tool name by ID.
pub fn find_tool_name_by_id(msg: &ParsedMessage, tool_use_id: &str) -> Option<String> {
    for tc in &msg.tool_calls {
        if tc.id == tool_use_id {
            return Some(tc.name.clone());
        }
    }
    if let (Some(src), Some(tur)) = (&msg.source_tool_use_id, &msg.tool_use_result) {
        if src == tool_use_id {
            if let Some(name) = tur.as_object().and_then(|o| o.get("toolName")).and_then(Value::as_str) {
                return Some(name.to_string());
            }
        }
    }
    None
}

/// Produces a short human-readable label for a tool_use call.
pub fn get_tool_summary(tool_name: &str, input: &Value) -> String {
    let get_str = |field: &str| -> String {
        input
            .as_object()
            .and_then(|o| o.get(field))
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string()
    };

    match tool_name {
        "Read" | "Edit" | "Write" => {
            let fp = get_str("file_path");
            if fp.is_empty() {
                return tool_name.to_string();
            }
            fp.rsplit(['/', '\\']).next().unwrap_or(&fp).to_string()
        }
        "Bash" => {
            let cmd = get_str("command");
            if cmd.is_empty() {
                return "shell command".to_string();
            }
            truncate_str(&cmd, 60)
        }
        "Grep" | "Glob" => {
            let pattern = get_str("pattern");
            if pattern.is_empty() {
                tool_name.to_string()
            } else {
                pattern
            }
        }
        _ => tool_name.to_string(),
    }
}

fn truncate_str(s: &str, max: usize) -> String {
    if s.len() <= max {
        return s.to_string();
    }
    let mut end = max;
    while !s.is_char_boundary(end) {
        end -= 1;
    }
    format!("{}...", &s[..end])
}

#[cfg(test)]
#[path = "extraction_tests.rs"]
mod tests;
